Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTP1Connections.swift

```diff
@@ -0,0 +1,438 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+
+extension HTTPConnectionPool {
+    struct HTTP1ConnectionState {
+        enum State {
+            /// the connection is creating a connection. Valid transitions are to: .backingOff, .available and .failed
+            case starting(retries: Int)
+            /// the connection is waiting to retry the establishing a connection. Valid transitions to: .starting and .closed
+            case backingOff(retries: Int)
+            case available(Connection, since: NIODeadline)
+            case leased(Connection)
+            case failed
+            case closed
+        }
+
+        private var state: State
+        private(set) var connectionID: Connection.ID
+        let eventLoop: EventLoop
+
+        init(connectionID: Connection.ID, eventLoop: EventLoop, retries: Int = 0) {
+            self.connectionID = connectionID
+            self.eventLoop = eventLoop
+            self.state = .starting(retries: retries)
+        }
+
+        var isConnecting: Bool {
+            switch self.state {
+            case .starting:
+                return true
+            case .backingOff, .failed, .closed, .available, .leased:
+                return false
+            }
+        }
+
+        var isBackingOff: Bool {
+            switch self.state {
+            case .backingOff:
+                return true
+            case .starting, .failed, .closed, .available, .leased:
+                return false
+            }
+        }
+
+        var isAvailable: Bool {
+            switch self.state {
+            case .available:
+                return true
+            case .backingOff, .starting, .leased, .failed, .closed:
+                return false
+            }
+        }
+
+        var isLeased: Bool {
+            switch self.state {
+            case .leased:
+                return true
+            case .backingOff, .starting, .available, .failed, .closed:
+                return false
+            }
+        }
+
+        var availableAndLastReturn: NIODeadline? {
+            switch self.state {
+            case .available(_, since: let lastReturn):
+                return lastReturn
+            case .backingOff, .starting, .leased, .failed, .closed:
+                return nil
+            }
+        }
+
+        mutating func started(_ connection: Connection) {
```

Comment:
I think this reads nicer as something thing `state.connectionStarted(connection)`. You're telling the state machine that something started, so tell it what.

Reply:
I guess normally I would agree. However we are down here to a single connection. and an invocation would therefore read like this `connectionState.connectionStarted(connectionBox)`. This is quite a number of connections in one line.